subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1235 @@
+
+	total_received += ctx_client->offset;
+	offset += ctx_client->offset;
+
+	if (offset >= HTTP_SERVER_MAX_RESPONSE_SIZE) {
+		offset = 0;
+	}
+
+	if (has_upgrade_header == false) {
+		const char *response = "HTTP/1.1 101 Switching Protocols\r\n"
+				       "Connection: Upgrade\r\n"
+				       "Upgrade: h2c\r\n"
+				       "\r\n";
+		if (sendall(ctx_client->client_fd, response, strlen(response)) < 0) {
+			close_client_connection(ctx_server, client_index);
+		}
+
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+		ctx_client->server_state = HTTP_SERVER_PREFACE_STATE;
+
+	} else {
+		data = content_200;
+		len = sizeof(content_200);
+
+		enum http_method method = parser.method;
+		const char *method_str = http_method_str(method);
+
+		LOG_DBG("HTTP Method: %s", method_str);
+
+		if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/results", 8) == 0) {
+			handle_get_request(ctx_server, ctx_client->client_fd);
+		} else if (strncmp(method_str, "POST", 4) == 0 &&
+			   strncmp(url_buffer, "/add", 4) == 0) {
+			char *json_start = strstr(ctx_client->buffer, "\r\n\r\n");
+
+			if (json_start) {
+				handle_post_request(ctx_server, json_start);
+			}
+
+			char response[] = "HTTP/1.1 200 OK\r\nContent-Type: "
+					  "text/plain\r\n\r\nAdded successfully.\n";
+			sendall(ctx_client->client_fd, response, strlen(response));
+		} else if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/", 1) == 0) {
+			sprintf(http_response,
+				"HTTP/1.1 200 OK\r\n"
+				"Content-Type: text/html\r\n"
+				"Content-Encoding: gzip\r\n"
+				"Content-Length: %d\r\n\r\n",
+				len);
+			if (sendall(ctx_client->client_fd, http_response, strlen(http_response)) <
+			    0) {
+				close_client_connection(ctx_server, client_index);
+			} else {
+
+				if (sendall(ctx_client->client_fd, data, len) < 0) {
+					LOG_ERR("sendall failed");
+					close_client_connection(ctx_server, client_index);
+				}
+			}
+		} else {
+			const char *not_found_response = "HTTP/1.1 404 Not Found\r\n"
+							 "Content-Length: 9\r\n\r\n"
+							 "Not Found";
+			if (sendall(ctx_client->client_fd, not_found_response,
+				    strlen(not_found_response)) < 0) {
+				close_client_connection(ctx_server, client_index);
+			}
+		}
+		close_client_connection(ctx_server, client_index);
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+	}
+
+	return 0;
+}
+
+int handle_http_done(struct http_server_ctx *ctx_server, struct http_client_ctx *ctx_client,
+		     int client_index)
+{
+	LOG_DBG("HTTP_SERVER_DONE_STATE");
+
+	close_client_connection(ctx_server, client_index);
+
+	return -errno;
+}
+
+int handle_http_frame_headers(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_SERVER_FRAME_HEADERS");
+
+	unsigned char response_headers_frame[16];
+
+	struct http_frame *frame = &ctx_client->current_frame;
+
+	print_http_frames(ctx_client);
+
+	if (ctx_client->offset < frame->length) {
```

Comment:
We cannot assume we'll fit entire frame payload into the buffer, this imposes a serious limitation (I remind you that the minimum frame size we have to support is 16k of bytes, and there's no way around it as it's the minimum specified by the RFC).

That's why I suggested that `http_hpack_parse_header` needs to be reworked (https://github.com/zephyrproject-rtos/zephyr/pull/59669#discussion_r1296070668). We should not be requesting the function to find a header in the buffer, instead, it should parse a single header (next in the buffer).

The function should:
 * return -EAGAIN if the header is not yet complete (for compressed headers it simple, as its a single byte only, but literal headers need to check for name length and value length fields to determine whether the header is complete)
 * if complete, return the header contents (strings) by filling user provided buffers (that would be the simplest approach). 
 * The function could return number of bytes parsed, or a negative error code in case of parsing errors. 
 
 And of course, once header has been parsed, we should "consume" the bytes from the buffer (`memmove()`) and decrease `current_frame->length` by the number of bytes consumed. Tracking `current_frame->length` allows us to detect, when there are no more headers to be expected (i. e. we've parsed the entire frame, `current_frame->length == 0`).

Reply:
> We cannot assume we'll fit entire frame payload into the buffer, this imposes a serious limitation (I remind you that the minimum frame size we have to support is 16k of bytes, and there's no way around it as it's the minimum specified by the RFC).

We might be able to get away with explicitly listing that as a deviation in e.g. docs, Kconfig. I would shy away from a runtime warning, but potentially a compile-time warning would be nice.

Nothing should prevent end users / system designers / product makers from configuring 16k on systems that have sufficient memory, but we can likely get away with far less.